### PR TITLE
New presentation - AWT

### PR DESCRIPTION
**Pull request by starsasumi**
#50

Use java.awt to implement a second presentation.

Note: I remove `PencilState` and use a delegate to "draw" pencils because the original `PencilState` could not fit this presentation.

### CODE REVIEW

**Comment by anqitong** [comment]
You have AWT and CircleAWT. The names make me think that CircleAWT should extends AWT. But here obviously CircleAWT extends ShapeState so maybe you should change the name or make CircleAWT extend AWT

**Comment by starsasumi** [comment]
@anqitong What about `AWT` -> `AWTFrame` and `CircleAWT` -> `AWTCircle`?

**Comment by anqitong** [comment]
I think it is better. At least it won't make people think that there is an extension between those two classes.

**Comment by starsasumi** [comment]
Thx, I will work on it later

**Comment by Thomas-dot-G** [comment]
Ok, I don't mind

**Comment by starsasumi** [comment]
@anqitong Classes renamed

File: files/svgAndAwt.svg [hunk 1]
```diff
@@ -0,0 +1,10 @@
+<?xml version="1.0" encoding="utf-8"?>
+<svg xmlns="http://www.w3.org/2000/svg" version="1.1" height="600" width="600">
+	<circle cx="30" cy="30" r="30" fill="rgb(0,0,255)" stroke="rgb(255,0,0)" stroke-width="4" />
+	<circle cx="100" cy="30" rx="40" ry="20" fill="rgb(0,0,255)" stroke="rgb(255,0,0)" stroke-width="4" />
+	<line x1="5" y1="60" x2="15" y2="100" stroke="rgb(255,0,0)" stroke-width="4" />
+	<polygon points=" 100,100 200,200 100,200" fill="rgb(0,0,255)" stroke="rgb(255,0,0)" stroke-width="4" />
+	<polyline points=" 250,100 100,300 20,300 250,200" stroke="rgb(255,0,0)" stroke-width="4" />
+	<rect x="300" y="50" width="150" height="100" fill="rgb(0,0,255)" stroke="rgb(255,0,0)" stroke-width="4" />
+	<text x="500" y="50" stroke="rgb(255,0,0)" stroke-width="4"">Hello, World!</text>
+</svg>
```
Comment (Thomas-dot-G) [inline files/svgAndAwt.svg]: what is that ?

File: files/test.svg [hunk 2]
```diff
@@ -0,0 +1,5 @@
+<?xml version="1.0" encoding="utf-8"?>
+<svg xmlns="http://www.w3.org/2000/svg" version="1.1" height="1024" width="1024">
+	<circle cx="50" cy="40" r="30" fill="rgb(0,0,255)" stroke="rgb(0,0,0)" stroke-width="4" />
+	<rect x="55" y="45" width="50" height="50" fill="rgb(0,255,0)" stroke="rgb(0,0,0)" stroke-width="4" />
+</svg>
```
Comment (Thomas-dot-G) [inline files/test.svg]: we don't put it here because it would make the first test fail

File: src/model/presentation/xml/StateDelegate.java [hunk 3]
```diff
@@ -0,0 +1,10 @@
+package model.presentation.xml;
+
+import model.Pencil;
+
+/**
+ * Created by Xiaxing SHI on 11/11/15.
+ */
+public interface StateDelegate {
+    String getPencilXML(Pencil p);
+}
```
Comment (Thomas-dot-G) [inline src/model/presentation/xml/StateDelegate.java:10]: Why is it necessary ? (I mean the whole change that comes with that)

**Comment by starsasumi** [comment]
Because getDrawing in ShapeState is generic, but getDrawing(PencilState) is heavily related to svg and just can not fit the case of awt. Also, awt presentation need to pass the `Graphics2D` object around, which is also not related to svg. So these (get xml from pencil, get `Graphics2D` object) are presentation related details and should not be share between presentations

File: src/model/presentation/xml/StateDelegate.java (hunk 3 above)
Comment (Thomas-dot-G) [inline src/model/presentation/xml/StateDelegate.java:10]: The pencil (crayon) is asked for every presentation by the subject

getDrawing(PencilState) can easily be used in awt, no ??

(The graphics2D object can be add in the awtDrawing

**Comment by starsasumi** [comment]
I fixed conflits. Can I merge?

File: src/model/presentation/awt/AWTCircle.java [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+        g.setStroke(new BasicStroke(this.circle.getPencil().getWidth()));
+        g.drawOval(this.circle.getCenter_x(), this.circle.getCenter_y(), this.circle.getRadius()*2, this.circle.getRadius()*2);
+
+        return this;
+    }
+
+    @Override
+    public void setShape(Shape shape) throws Exception {
+        if(shape instanceof Circle){
+            this.circle = (Circle) shape;
```
Comment (anqitong) [inline src/model/presentation/awt/AWTCircle.java:38]: I was wondering if we can do something better than (Circle) circle which is coercion. What about using a conversion method, something like that : 
if (shape instanceof Circle) return shape.convertToCircle();

**Comment by starsasumi** [comment]
@anqitong There's no different, since shape is an instance of Circle, there is nothing to convert (in the convertToCircle method we still need to cast itself to circle and return). In fact I prefer to remove this setShape method because a ShapeState is something too lightweight and we would just use an object one time and then throw it away. So we don't need to call the setShape method.

**Comment by anqitong** [comment]
Ok, then it's fine with me

**Comment by Thomas-dot-G** [comment]
I got very strange errors when running it:
`(eog:23372): GLib-WARNING **: GError set over the top of a previous GError or uninitialized memory.
This indicates a bug in someone's code. You must ensure an error is NULL before it's set.
The overwriting error message was: Error domain 1 code 73 on line 9 column 31 of file:///home/thomas/Documents/Programs/LangageDessinVectoriel/files/svgAndAwt.svg: Couldn't find end of Start Tag text`

(And the drawing is a little strange^^)

File: src/model/presentation/awt/AWTFrame.java [hunk 5]
```diff
@@ -0,0 +1,124 @@
...
+                state = new AWTCircle((Circle)shape, this);
+                break;
+            case Ellipse:
+                state = new AWTEllipse((Ellipse)shape, this);
+                break;
+            case Line:
+                state = new AWTLine((Line)shape, this);
+                break;
+            case Path:
+//                state = new PathXML((Path)shape);
```
Comment (Thomas-dot-G) [inline src/model/presentation/awt/AWTFrame.java]: Add a TODO here maybe

File: src/model/presentation/awt/AWTRectangle.java [hunk 6]
```diff
@@ -0,0 +1,40 @@
...
+
+        g.setColor(this.rectangle.getPencil().getColor());
+        g.setStroke(new BasicStroke(this.rectangle.getPencil().getWidth()));
+        g.drawRect(this.rectangle.getX(), this.rectangle.getY(), this.rectangle.getWidth(), this.rectangle.getHeight());
+
+        return this;
+    }
+
+    @Override
+    public void setShape(Shape shape) throws Exception {
```
Comment (Thomas-dot-G) [inline src/model/presentation/awt/AWTRectangle.java:37]: All the setShape can be done I think

File: src/model/presentation/xml/SVGFile.java [hunk 7]
```diff
@@ -219,4 +217,9 @@ public SVGFile clone(){
 		
 	}
 
+	@Override
+	public String getPencilXML(Pencil p) {
```
Comment (Thomas-dot-G) [inline src/model/presentation/xml/SVGFile.java:221]: I still think that putting the pencil here rather that like before in a separate class is not necessary, and even a little odd...

**Comment by starsasumi** [comment]
@Thomas-dot-G The error is related to SVG? I have no error about the AWT, but I do get this in the SVG version output:

screen shot 2015-11-14 at 15 23 19

I think that's a bug in our xml implementation. For the other parts of codes, I do it now.

**Comment by Thomas-dot-G** [comment]
Yes exactly, but as you changed the xml also, it might come from that.

**Comment by Thomas-dot-G** [comment]
It should be compatible.
And can you add some comments also in the main and a test (can be the same as the main)

**Comment by starsasumi** [comment]
@Thomas-dot-G I found the bug, there was an extract `"` in the TextXML codes. It should be fine now ;)

**Comment by starsasumi** [comment]
For adding a JUnit test, I tried it once, but Junit will kill all JFrame immediately after test codes are finished. Do you have any idea how we make Junit not to kill JFrames?

**Comment by Thomas-dot-G** [comment]
No, not for JFrames...
You can just create the object, and then do some assert on the element inside the object to see if they are well here... (some small JUnits, not really on the display...)
But, I don't know otherwise

**Comment by Thomas-dot-G** [comment]
Now xml works, but on the xml, I can't see the ellipse, and the hello world is not really readable...
And the polyline is filled in the xml...

**Comment by starsasumi** [comment]
I made the DrawAwt now as a test. For xml, I confirm the problems. But I prefer to open another issue to fix the xml implementation (Since I have not work a lot on SVG and I didn't modify anything about xml outputs). Ok?

**Comment by Thomas-dot-G** [comment]
Ok, fine with me then

**Pull request merged**